common/flatpak-utils.c

```diff
@@ -105,12 +105,17 @@ flatpak_error_quark (void)
 gboolean
 flatpak_fail_error (GError **error, FlatpakError code, const char *fmt, ...)
 {
+  GError *new;
+
   if (error == NULL)
     return FALSE;
 
   va_list args;
   va_start (args, fmt);
-  GError *new = g_error_new_valist (FLATPAK_ERROR, code, fmt, args);
+  if (g_error_matches (*error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
```

Comment:
This seems wrong – in this function, surely we expect `error == NULL || *error == NULL`? `g_propagate_error` internally makes that assertion.

Reply:
hmm tricky. Well `*error` can be NULL. See how `flatpak_fail_error` is getting used [here](https://github.com/flatpak/flatpak/blob/master/common/flatpak-dir.c#L11280-L11282). 

OTOH, it will be NULL-dereferencing if we check for IO_ERROR_CANCELLED. 
Maybe what we really need is :
`if ((*error != NULL) && g_error_matches (*error, G_IO_ERROR, G_IO_ERROR_CANCELLED))`

Reply:
My point is that calling `g_propagate_error (error, ...)` is a bug in the case where `error != NULL && *error != NULL`. So, this function implicitly makes the same assumption: `error == NULL || *error == NULL`.